includes/admin/class-amp-cache-utilities.php

```diff
@@ -0,0 +1,251 @@
+	 * @param  string 	$scheme       	Optional. 'http' or 'https' If provided, this overrides the
+	 *                                	scheme in the URL. If null, scheme is taken from the URL.
+	 * @return string/bool              The AMP cache URL on success, false on failure.
+	 */
+	public static function get_amp_cache_url_for_resource( $url, $content_type, $scheme = null ) {
+		$amp_cache_resource_path = self::get_amp_cache_path_for_url( $url, $content_type );
+		if ( $amp_cache_resource_path ) {
+			return self::$amp_cache_url_base . '/' . ltrim( $amp_cache_resource_path, '/' );
+		}
+		return false;
+	}
+
+	/**
+	 * Given any URL and content type, calculate the AMP cache update URL for this resource.
+	 * 
+	 * @param  string 	$url     		The url to calculate the AMP cache update URL from.
+	 * @param  string 	$content_type 	Currently only supports 'c' for posts and 'i' for images.
+	 * @param  string 	$scheme       	Optional. 'http' or 'https' If provided, this overrides the
+	 *                                	scheme in the URL. If null, scheme is taken from the URL.
+	 * @return string/bool              The AMP cache udpate URL on success, false on failure.
+	 */
+	public static function get_amp_cache_update_url_for_resource( $url, $content_type, $scheme = null ) {
+		$amp_cache_resource_path = self::get_amp_cache_path_for_url( $url, $content_type );
+		if ( $amp_cache_resource_path ) {
+			return self::$amp_cache_update_url_base . '/' . ltrim( $amp_cache_resource_path, '/' );
+		}
+		return false;
+	}
+
+	/**
+	 * Given a post_id or WP_Post onbject, calculate the AMP cache path part (the part after 
+	 * https://cdn.ampproject.org) for this post.
+	 * 
+	 * @param  int/WP_Post	$post 			The post to get rhe AMP cache path for.
+	 * @param  string 		$content_type 	Currently only supports 'c' for posts and 'i' for images.
+	 *                                 		Optional, if provided, this overrides the actual content
+	 *                                 		type of the post. If null, the content type is determined
+	 *                                 		by post_type.
+	 * @param  string 	$scheme       		Optional. 'http' or 'https' If provided, this overrides the
+	 *                                		scheme in the URL. If null, scheme is taken from the URL.
+	 * @return string/bool              	The AMP cache path part on success, false on failure.
+	 */
+	public static function get_amp_cache_path_for_post( $post, $content_type = null, $scheme = null ) {
+		$permalink = get_permalink( $post );
+
+		// If permalink couldn't be retrieved, return failure.
+		if ( false === $permalink ) {
+			return false;
+		}
+
+		// determine $content_type, if not specified
+		if ( null == $content_type ) {
+			$post_type = get_post_type( $post );
+			switch ( $post_type ) {
+				case 'attachment':
+					$content_type = 'i';
+					break;
+				case 'post':
+					$content_type = 'c';
```

Comment:
I was thought about this too and was wondering if I should set up a filter to allow only certain custom post types to be included since some custom post types might not be appropriate for the AMP cache.